### ISSUE TITLE
Errno 99 Cannot assign requested address when running scapy in namespace

### ISSUE THREAD

**Issue by twix1312** [secdev/scapy#938]
I'm trying to run a simple scapy application (test.py) that sends an ICMP packet. 

```
from scapy.all import send, IP, ICMP
send(IP(src="192.168.121.75", dst="192.168.251.62")/ICMP())
```

I run this command in a network namespace, as follows:

`ip netns exec nic_1_ns python test.py`

And get the following error:

```
Traceback (most recent call last):
  File "gtpTest.py", line 2, in <module>
    from scapy.all import send, IP, ICMP
  File "/usr/lib/python2.7/site-packages/scapy/all.py", line 25, in <module>
    from scapy.route import *
  File "/usr/lib/python2.7/site-packages/scapy/route.py", line 163, in <module>
    conf.route=Route()
  File "/usr/lib/python2.7/site-packages/scapy/route.py", line 22, in __init__
    self.resync()
  File "/usr/lib/python2.7/site-packages/scapy/route.py", line 31, in resync
    self.routes = read_routes()
  File "/usr/lib/python2.7/site-packages/scapy/arch/linux.py", line 177, in read_routes
    ifreq = ioctl(s, SIOCGIFADDR,struct.pack("16s16x",LOOPBACK_NAME))
IOError: [Errno 99] Cannot assign requested address

```

Details of the namespace ip configuration is:
```
# sudo ip netns exec nic_1_ns ifconfig
enp5s0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.121.75  netmask 255.255.255.0  broadcast 192.168.121.255
        inet6 fe80::f27d:68ff:feb8:c0fb  prefixlen 64  scopeid 0x20<link>
        ether f0:7d:68:b8:c0:fb  txqueuelen 1000  (Ethernet)
        RX packets 57071  bytes 7852014 (7.4 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1061  bytes 109367 (106.8 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

enp5s0:0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.121.90  netmask 255.255.255.0  broadcast 192.168.121.255
        ether f0:7d:68:b8:c0:fb  txqueuelen 1000  (Ethernet)
```

I'm using:
**Python 2.7.5**
**scapy (2.3.3)**
**CentOS 7.3**

If I run the command outside the namespace, the code works fine. 
It seems that ioctl command is failing and it can't bind to the loopback IP address. Ideally, it should try to run this command within the namespace itself. Is there any configuration file where I can add this namespace specific info? Is there any workaround for this or does scapy not have support to work within network namespaces?

**Comment by speakinghedge** [comment]
try to set the loopback device up before calling your script:
```
sudo ip netns exec nic_1_ns ip link set lo up
```

**Comment by speakinghedge** [comment]
(it was not shown in the output of your call to ifconfig and that indicates the interface is down. you can force non active interfaces to show up by appending option -a --or-- use ip link to get the list & state of all your interfaces)

**Comment by twix1312** [comment]
After enabling the loopback interface, I can see that the program outputs "Sent 1 packet". However, this packet isn't actually being sent across the network (didnt show up in the wire shark capture).

Why does the loop back interface need to be up and running for this to work anyway? It doesn't make sense to me. Could you please elaborate?

**Comment by speakinghedge** [comment]
> After enabling the loopback interface, I can see that the program outputs "Sent 1 packet". However, this packet isn't actually being sent across the network (didnt show up in the wire shark capture).

as you run the script in a namespace I guess there is no route available. what is the output of 

```
sudo ip netns exec nic_1_ns ip route
```

(there is usually no default route added so you may end up with a single entry to 192.168.121.0/24 - but you try to send the message to a host in another sub net 192.168.**251**.0/24)

> Why does the loop back interface need to be up and running for this to work anyway? It doesn't make sense to me. Could you please elaborate?

during the bootstrap of Scapy a table containing all available routes is created. this table also includes the loopback device - the error message will guide you to the relevant piece of code:
```
File "/usr/lib/python2.7/site-packages/scapy/arch/linux.py", line 177, in read_routes
```

if the loopback was never up before no address (usually 127.0.0.1) was assigned. SIOCGIFADDR tries to return an INET address and will fail if no address is available. therefore the ioctl fails if the namespace was created without bringing up the loopback.

**Comment by twix1312** [comment]
So the loopback address makes sense now, thank you.

The gateways to which the NIC card's are connected have interVLAN routing, so there is no need to add explicit routes. Nevertheless:

```
sudo ip netns exec nic_1_ns ip route
default dev enp5s0  scope link
192.168.121.0/24 dev enp5s0  proto kernel  scope link  src 192.168.121.75
```

To prove that packet's are routable, a simple ping test works:

```
ip netns exec nic_1_ns ping 192.168.252.62
PING 192.168.252.62 (192.168.252.62) 56(84) bytes of data.
64 bytes from 192.168.252.62: icmp_seq=2 ttl=63 time=0.469 ms
64 bytes from 192.168.252.62: icmp_seq=3 ttl=63 time=0.295 ms
64 bytes from 192.168.252.62: icmp_seq=4 ttl=63 time=0.432 ms

```


So this now raises the question that when the scapy program sent the packets, why weren't they picked up in the wireshark trace?

**Comment by speakinghedge** [comment]
To be able to (try to) help you I need to know a little more about the structure of your setup
(network structure, point used to capture traffic).

side quest:
where is the route for enp5s0:0? I guess this is your tagged device and should also appear in the routing table.

**Comment by guedou** [comment]
Closed by #958
